### ISSUE TITLE
Windows Defender causing slowdown on contents of .cargo\bin folder

### ISSUE THREAD

**Issue by krolli** [rust-lang/cargo#5028]
Hi,
a recently I noticed that running rustfmt is much slower than it used to be. Digging into it, I found out the slowdown is caused by Windows Defender (specifically MsMpEng.exe, Antimalware Service Executable), but is not limited to rustfmt. In fact, rustfmt in toolchain runs fast on its own, it's rustfmt binary in .cargo\bin directory that is causing large CPU usage spikes in scanning process. Setting Windows Defender to ignore rust-related directories or changing PATH environment variable to search directly in toolchain directory instead of taking the roundabout way through .cargo/bin dispatch binaries fixes the problem.

Since this affects all binaries, it also drags build performance down. I did some measurements on my test project (relatively small) with/without exception in Windows Defender and with direct path to toolchain and indirect path via .cargo:

full rebuild:
no exception, indirect path : 143s
exception, indirect path : 92s
no exception, direct path: 84s
exception, direct path: 83s

incremental build (single changed file):
no exception, indirect path: 4.6s (11.6s from button press)
exception, direct path: 2.8s (3.8s from button press)

Incremental build times are combination of cargo output value (shorter) and ETW trace measurement from pressing build button until link.exe is finished. Incremental build also includes running rustfmt (since the file gets saved).

**Comment by ehuss** [comment]
I have noticed this, too.  The rustup wrapper itself adds a significant amount of time to execution.

Exe | With Defender | Without Defender
----|---------------|-----------------
rustc -V | 0.7s | 0.09s
~/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/rustc.exe -V | 0.1s | 0.08s
cl.exe | 0.08s  | 0.056s
MinGW gcc --version | 0.2s | 0.11s

This may not be a cargo-specific issue.  It would be nice if the rustup wrapper didn't add a nearly 8x slowdown.

I typically just turn it off while doing development.  I don't know if there is a realistic workaround.

**Comment by krolli** [comment]
I thought this was a cargo thing, given ".cargo" folder name.

**Comment by alexcrichton** [comment]
Oh dear this seems bad! Is there a known way to fix this though? I just realized that I ended up disabling Windows Defender for the .cargo/.rustup directories presumably long ago...

**Comment by krolli** [comment]
I just heard from some people this happens even for nvidia gl dll, which is definitely signed.

I don't know of any way to convince Windows Defender to trust those binaries (short of contacting Microsoft). Workaround I am using is I just added path to toolchain directly into PATH variable, ahead of .cargo\bin, but I'm not sure why those dispatching binaries are there, so maybe this is not acceptable in general. Switching toolchains in rustup would then be just changing PATH variable. Cleaner way to do that might be to introduce another variable (eg. RUST_TOOLCHAIN) and just reference that from PATH using.

But, like I said, I'm not sure what is being solved by those dispatch binaries, so this might not be the right way.

**Comment by steffengy** [comment]
@krolli 
Is this related to the referenced rustup issue?
Try replacing your `rustc.exe`, `rustup.exe` or whatever slow rust binary you're using with https://win.rustup.rs/x86_64.

**Comment by krolli** [comment]
@steffengy 
It is indeed faster. While adding exception to Defender and directly invoking rust tools still shaves off some time, it is not nearly as pronounced as before. I did a quick trace and it appears that scanning still occurs. It seems to take up about quarter of the original time, which is somewhat better improvement than I'd expect if the slowdown was linear in size of scanned binaries. Still, I think speedup is largely due to their smaller size.

**Comment by steffengy** [comment]
@krolli 
> Still, I think speedup is largely due to their smaller size.

It actually isn't. 
i686-msvc is still slow, even though it's only 5MB.
x86-64-gnu (13 MB) is fast, even though the file size is even bigger.
So it's likely related to some architectural differences (and special treatment from windows defender).

Anyways those fixes landed in rustup.

**Comment by krolli** [comment]
@steffengy 
I haven't tried i686-msvc, but I though I had x86-64-gnu installed before when it was really slow. I'll play around with it later to find out more.

**Comment by krolli** [comment]
@steffengy 
You are right. Both gnu and msvc 32-bit binaries are taking way longer to scan.

**Comment by retep998** [comment]
>Is this still relevant?

Yes.

>If so, what is blocking it?

The lack of a developer willing to put in the time and effort to get this fixed.

>Is it known what could be done to help move this forward?

Someone could use a tool like procmon to compare how slow various syscalls are with and without Defender. A side by side comparison should make it clear which syscalls are impacted most by Defender and then the code can be adjusted to avoid those syscalls.

**Comment by weihanglo** [comment]
FWIW, <https://github.com/rust-lang/cargo/pull/11917> might speed up this a bit with <https://github.com/rust-lang/rustup/pull/3178> enabled.

@rustbot label +S-needs-design
